File: BOJ/[G4]2473 세 용액.cpp
// [G4]2473 세 용액
// https://www.acmicpc.net/problem/2473

#include<cstdio>
#include<algorithm>
#define ABS(A) ((A) > 0 ? (A) : -(A))
#define EXIT printf("%d %d %d",a,b,c)
#pragma warning(disable: 4996)
using namespace std;

long long arr[5000];
int n, a, b, c;
long long _min = 3e9;
int main() {
	scanf("%d", &n);
	for (int i = 0; i < n; i++)
		scanf("%lld", &arr[i]);
	
	sort(&arr[0], &arr[n]);
	for (auto i = &arr[0]; i != &arr[n - 2]; i++){
		for (auto f = i + 1, r = &arr[n - 1]; f != r;) {
			long long sum = (*i) + (*f) + (*r);
			if (_min > ABS(sum)) {
				_min = ABS(sum);
				a = *i, b = *f, c = *r;
			}
			if (sum < 0)f++;
			else if (sum > 0) r--;
			else { EXIT; return 0; }
		}
	}
	EXIT;
}

File: BOJ/[G4]2473 세 용액.py
# [G4]2473 세 용액
# https://www.acmicpc.net/problem/2473

# p1고정 후 p2, p3 움직임

import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
arr.sort()

max = int(3e9) + 1
temp = []
for p1 in range(0, n-1):
    p2, p3 = p1 + 1, n - 1
    while p2 != p3:
        x = arr[p1] + arr[p2] + arr[p3]
        if x == 0:
            print(arr[p1], arr[p2], arr[p3])
            sys.exit()

        if abs(x) < max:
            temp = [arr[p1],arr[p2],arr[p3]]
            max = abs(x)

        if x < 0:
            p2 += 1
        else:
            p3 -= 1

print(*temp)